src/pytorch_lightning/trainer/connectors/checkpoint_connector.py

```diff
@@ -209,6 +209,10 @@ def restore_training_state(self) -> None:
     def restore_precision_plugin_state(self) -> None:
         """Restore the precision plugin state from the pre-loaded checkpoint."""
         prec_plugin = self.trainer.precision_plugin
+
+        if isinstance(prec_plugin, ApexMixedPrecisionPlugin) and self.trainer.state.fn == TrainerFn.FITTING:
```

Comment:
Why not raise this error within the `ApexMixedPrecisionPlugin`? That way it would be encapsulated within the responsible plugin.

Reply:
Because inside there, we don't know whether the trainer is training or testing. 
I don't know if I can move it inside like you said. There would still have to be an isinstance check somewhere to set this special attribute.

Reply:
I imagined it to be something like this:


```python
def load_state_dict(self, state_dict):
    self._state_dict_loaded = True
    return super().load_state_dict(state_dict)

def optimizer_step(self, *args, **kwargs): # args and kwargs for simplicity, ofc keep real signature
    # in here we definitely know, that we are training
   if self._state_dict_loaded:
        raise MisconfigurationException(...)
   ... # continue with usual optimizer step.

```

Wouldn't that work?

Reply:
I understand what you mean now. Yes, I definitely like this more!